### ISSUE TITLE
TypeScript error when setting the `module` option to `esnext`

### ISSUE THREAD

**Issue by CarsonMcKinstry** [dotansimha/graphql-code-generator#5757]
**Describe the bug**
 

It appears that if you set the `compilerOptions.module` setting in a `tsconfig.json` to `esnext`, this causes `graphql-gen` to fail with the following error:

```
Unable to load from file "path/to/module.ts": Cannot use import statement outside a module
```

This would be useful for me, as I am working the TypeScript in a Next.js project and the framework automatically sets this option to `esnext` for me. 

**To Reproduce**

[Here's a reproduction instead](https://codesandbox.io/s/hopeful-tdd-hlbfc?file=/tsconfig.json)

**Comment by CarsonMcKinstry** [comment]
Did a bit more digging, if you set the `module` flag to _anything_ it will fail.

**Comment by ardatan** [comment]
I don't think this is related to codegen. Codegen just creates TypeScript files like you write. For example in the repo you provided, TS-Node needs `commonjs` because NodeJS uses CommonJS. It would still fail even if you don't use codegen there.

**Comment by CarsonMcKinstry** [comment]
Here is a repo that produces the bug:

https://github.com/CarsonMcKinstry/nextjs-graphqlgen-bug

**Comment by ardatan** [comment]
@CarsonMcKinstry I don't see any error when I navigate to `/api/graphql`. Could you help me reproduce it?
https://codesandbox.io/s/relaxed-dijkstra-yflms?file=/src/pages/api/graphql.ts

**Comment by CarsonMcKinstry** [comment]
So that's not exactly the issue. Running the next server works just fine. Generating the typescript types is the problem. In the repo I provided, running `yarn generate` produces the error.

**Comment by ardatan** [comment]
@CarsonMcKinstry Because it tries to load TS file in Codegen with TSNode and since CommonJS doesn't support ESM import syntax, it fails. You can have a js file that loads ts-node/register with custom compiler options that forces commonjs output. 
But I'd recommend to point to the individual files that contains type definitions directly because the code file loaders normally try to pluck type definitions from the raw code then try to load it as a module. So having a specific glob that points to the type definitions would be better option for you.
`schema: src/graphql/modules/**/typeDefs.ts`

**Comment by CarsonMcKinstry** [comment]
Ah, damn. Looks like you're right. 

With that being said, is there a way to pass flags to the `require` command? Perhaps this could be a way to tell `ts-node` to compile using es modules.

**Comment by chucksellick** [comment]
This could be solved if the plugin has an option to load tsconfig from a different filename. I already use this in my project, some packages have a tsconfig.dist.json which has slightly different options for when I'm building a distribution vs just running typescript direct in node. I would like to have a tsconfig.codegen.json to not conflict with options in my tsconfig.json which is used for building a React app. (I am having the same issue with create-react-app rather than next.js but same thing, it keeps resetting the config every time I run a build).

**Comment by FeliciaAKelley** [comment]
Use heroku for deployment it helps 

Sent from Yahoo Mail on Android 
 
  On Fri, Jun 18, 2021 at 10:20 AM, Peter ***@***.***> wrote:   


This could be solved if the plugin has an option to load tsconfig from a different filename. I already use this in my project, some packages have a tsconfig.dist.json which has slightly different options for when I'm building a distribution vs just running typescript direct in node. I would like to have a tsconfig.codegen.json to not conflict with options in my tsconfig.json which is used for building a React app. (I am having the same issue with create-react-app rather than next.js but same thing, it keeps resetting the config every time I run a build).

**Comment by niranjangajare** [comment]
I Found a quick solution for switching the module in `tsconfig.json` when using Next.js

```jsonc
// tsconfig.json
{
  "ts-node": {
    // these options are overrides used only by ts-node
    "compilerOptions": {
      "target": "ES5",
      "module": "CommonJS"
    }
  },
  "compilerOptions": {
     ...
  }
}
```

Also if you are using `"paths"` in your `"compilerOptions"`, you will need to install [`tsconfig-paths`](https://www.npmjs.com/package/tsconfig-paths) package and update your `require` field in `codegen.yaml` like this

```yaml
# codegen.yaml
require:
  - ts-node/register
  - tsconfig-paths/register
```

**Comment by huw** [comment]
This is fixable by running `NODE_OPTIONS='--loader ts-node/esm --experimental-specifier-resolution=node' graphql-codegen`. With this setup, you don’t need to specify a `require` option in your `codegen.yml` file either (doing so won’t work regardless since the CLI can’t `import()` `ts-node/esm`).

**Comment by locona** [comment]
I ran into this problem when I upgraded from graphql-codegen v2 to v3.
Since it works fine in v2, is there something wrong with it from v3?

```
NODE_OPTIONS='--loader ts-node/esm --experimental-specifier-resolution=node' npx graphql-codegen --config codegen.yml
```
